sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/StructuredMessageEncoder.java

```diff
@@ -0,0 +1,359 @@
+        this.contentLength = contentLength;
+        this.flags = flags;
+        this.segmentSize = segmentSize;
+        this.numSegments = Math.max(1, (int) Math.ceil((double) this.contentLength / this.segmentSize));
+        this.messageLength = calculateMessageLength();
+        this.contentOffset = 0;
+        this.currentSegmentNumber = 0;
+        this.currentRegion = SMRegion.MESSAGE_HEADER;
+        this.currentRegionLength = getMessageHeaderLength();
+        this.currentRegionOffset = 0;
+        this.checksumOffset = 0;
+        this.messageCRC64 = 0;
+        this.segmentCRC64s = new HashMap<>();
+    }
+
+    private int getMessageHeaderLength() {
+        return V1_HEADER_LENGTH;
+    }
+
+    private int getSegmentHeaderLength() {
+        return V1_SEGMENT_HEADER_LENGTH;
+    }
+
+    private int getSegmentFooterLength() {
+        return (flags == Flags.STORAGE_CRC64) ? CRC64_LENGTH : 0;
+    }
+
+    private int getMessageFooterLength() {
+        return (flags == Flags.STORAGE_CRC64) ? CRC64_LENGTH : 0;
+    }
+
+    private byte[] generateMessageHeader(int version, int size, Flags flags, int numSegments) {
+        // 1 byte version, 8 byte size, 2 byte flags, 2 byte numSegments
+        ByteBuffer buffer = ByteBuffer.allocate(13).order(ByteOrder.LITTLE_ENDIAN);
+        buffer.put((byte) version);
+        buffer.putLong(size);
+        buffer.putShort((short) flags.getValue());
+        buffer.putShort((short) numSegments);
+
+        return buffer.array();
+    }
+
+    private byte[] generateSegmentHeader(int number, int size) {
+        // 2 byte number, 8 byte size
+        ByteBuffer buffer = ByteBuffer.allocate(10).order(ByteOrder.LITTLE_ENDIAN);
+        buffer.putShort((short) number);
+        buffer.putLong(size);
+
+        return buffer.array();
+    }
+
+    private void updateCurrentRegionLength() {
+        switch (currentRegion) {
+            case MESSAGE_HEADER:
+                currentRegionLength = getMessageHeaderLength();
+                break;
+
+            case SEGMENT_HEADER:
+                currentRegionLength = getSegmentHeaderLength();
+                break;
+
+            case SEGMENT_CONTENT:
+                // last segment size is remaining content
+                if (currentSegmentNumber == numSegments) {
+                    currentRegionLength = contentLength - ((currentSegmentNumber - 1) * segmentSize);
+                } else {
+                    currentRegionLength = segmentSize;
+                }
+                break;
+
+            case SEGMENT_FOOTER:
+                currentRegionLength = getSegmentFooterLength();
+                break;
+
+            case MESSAGE_FOOTER:
+                currentRegionLength = getMessageFooterLength();
+                break;
+
+            default:
+                throw new IllegalArgumentException("Invalid SMRegion " + currentRegion);
+        }
+    }
+
+    private int tell() {
```

Comment:
Instead of needing to implement all of this mapping logic, could we just keep track of how much encoded data we've returned? It looks like `tell()` really is only used to compare when we've reached the end of the expected message length. So, if we keep a running total of how much we've encoded, that would probably be easier to follow/track.

I think we would only need this logic if we were implementing our own readable stream, which we have not decided on, where we could seek out of band of reading data.